torch/csrc/autograd/custom_function.h

```diff
@@ -11,4 +11,49 @@ TORCH_API variable_list _wrap_outputs(
   const std::unordered_set<at::TensorImpl*> &dirty_inputs,
   const at::ArrayRef<Variable> raw_outputs,
   const std::shared_ptr<Function> &cdata);
+
+// To use custom autograd operations implement a CFunction subclass with
+// static backward and forward functions
+//
+// forward() can take as many arguments as you want and should return a
+// variable list. Use of any direct Variable arguments will be registered in
+// the graph but no vectors/sets or any other data structures will be traversed.
+//
+// backward() will be given a variable list containing as many Variables as
+// there were outputs from forward. It should return as many Variables as there
+// were inputs with each of them containing the gradient w.r.t. its
+// corresponding input
+//
+// For example:
+// class MyFunction : public CFunction<MyFunction> {
+//   public:
+//   static variable_list forward(AutogradContext *ctx, int n, Variable var);
+//
+//   static variable_list backward(AutogradContext *ctx, variable_list grad_output);
+// };
+// To use MyFunction
+// Variable x;
+// MyFunction::apply(6, x);
+
+template <class T>
+struct CFunction {
```

Comment:
naming bikeshedding: since it's called `torch.autograd.Function` in Python, it would be best if we could similarly expose this as `torch::autograd::Function`. Unfortunately, that name is already taken here. I wonder if some renaming isn't in order...

Reply:
Not sure about the naming here as well.
In cpp, the current `Function` (and it's subclasses) are nodes in the graph we build.
In python, `Function` can be seen as the definition of an autograd atomic operation.
The problem is that both meanings are quite deeply rooted now no?

But I agree that in cpp, the current `Function` should become `Node` (and `PyNode`, `CppNode`etc) and this class should become `Function`.
Would it be ok for backward compatibility?

Reply:
Renaming in https://github.com/pytorch/pytorch/pull/23269